### ISSUE TITLE
Cannot update entity with relation

### ISSUE THREAD

**Issue by olingern** [typeorm/typeorm#4477]
**Issue type:**

[ ] question
[x ] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[ ] `mysql` / `mariadb`
[ ] `oracle`
[x] `postgres`
[ ] `cockroachdb`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[x ] `latest`
[ ] `@next`
[ ] `0.x.x` (or put your version here)

### Example

Given these entities:

```typescript
@Entity('users')
export default class UserEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;
    
  @Column()
  name: string

  @OneToMany(type => Photo, photo => photo.user, { eager: true })
  photos: Photo[];
}


@Entity('photos')
export class Photo {

  @PrimaryGeneratedColumn()
  id: number;

  @Column()
  url: string;

  @ManyToOne(type => User, user => user.photos)
  user: User;
}
```

Given these update and create functions:

```typescript

export async function createUser(user): Promise<UserEntity> {
  const userRepo: Repository<UserEntity> = getRepository(UserEntity);

  const newUser = userRepo.create(user);
  return userRepo.save(newUser)
}

export async function updateUser(existingUser: UserEntity) {
  const userRepo: Repository<UserEntity> = getRepository(UserEntity);
	
  return userRepo.update({ id: existingUser.id }, existingUser)
}

```

Updating fails:

```typescript
const user =  await createUser({ name: 'foo'})
user.name = 'bar'
await updateUser(user)
```

With this error:
`EntityColumnNotFound: No entity column "photos" was found.`

**Comment by olingern** [comment]
It seems this is the source of the error:
https://github.com/typeorm/typeorm/blob/7d7fd9f00ca45013d86ba8d570b66bd15b62d69d/src/query-builder/UpdateQueryBuilder.ts#L373

It's expected that `photos` is a column, but it is not.

If I comment the `throw` out, it executes fine. Logging the metadata I see:

```
ownRelations: [
  RelationMetadata {
    isTreeParent: false,
    isTreeChildren: false,
    isPrimary: false,
    isLazy: false,
    isEager: true,
    persistenceEnabled: true,
    isCascadeInsert: true,
    isCascadeUpdate: true,
    isCascadeRemove: true,
    isNullable: true,
    isOwning: false,
    isOneToOne: false,
    isOneToOneOwner: false,
    isWithJoinColumn: false,
    isOneToOneNotOwner: false,
    isOneToMany: true,
    isManyToOne: false,
    isManyToMany: false,
    isManyToManyOwner: false,
    isManyToManyNotOwner: false,
    foreignKeys: [],
    joinColumns: [],
    inverseJoinColumns: [],
    entityMetadata: [Circular],
    embeddedMetadata: undefined,
    target: [Function: UserEntity],
    propertyName: 'photos',
    relationType: 'one-to-many',
    givenInverseSidePropertyFactory: [Function],
    onDelete: undefined,
    onUpdate: undefined,
    deferrable: undefined,
    type: [Function: OrganizationEntity],
    propertyPath: 'photos',
    inverseEntityMetadata: [EntityMetadata],
    inverseSidePropertyPath: 'user',
    inverseRelation: [RelationMetadata]
  }
],
```

Just a question. Why is it necessary to throw when a relation column can't be found? Those entities need to be updated and saved on their own, no? 

I'm relatively new to typeorm, so I could just be doing something wrong.

**Comment by pbirsinger** [comment]
Yea this is pretty annoying. Happening to me too.

**Comment by a4xapps** [comment]
yeah same with me, for some reason this error doesn't happen with my ManyToMany relation. I wonder if the @JoinTable() has something to do with that.

but looking further into it, it looks as if the photo needs to be added to the table through its own post request and not through the user update request... weird.


Trying this:
![image](https://user-images.githubusercontent.com/52838454/62151562-8ff95500-b2ce-11e9-8db8-cb50635df24e.png)

then get this error: EntityColumnNotFound: No entity column "photos" was found.

I am new to typeorm as well.

**Comment by lastch** [comment]
any solution on this?

**Comment by justspamjustin** [comment]
Is there a solution to this issue? I'm also running into this problem.

**Comment by justspamjustin** [comment]
I was able to fix this issue by using the `save` method rather than the `update` method.

**Comment by Tidiane0843** [comment]
> I was able to fix this issue by using the `save` method rather than the `update` method.

When I used save method, In the User table, my user(who I want to update) got a new ID and in the table Photo, the userID is set to null for the photos which were linked to my user. 
Is is the same for you?

**Comment by mehtamanan** [comment]
@Tidiane0843 This may help #4122

**Comment by Tidiane0843** [comment]
@mehtamanan thanks. I managed to update data using `save` method yesterday. 
I just have to change the decorator `@ManyToOne` in the photo entity like this 

```
@ManyToOne(type => User, user=> user.photo, 
{onUpdate: 'CASCADE', onDelete: 'CASCADE'}
)
```

**Comment by mehtamanan** [comment]
@Tidiane0843 Wut! onUpdate sets the sql ON UPDATE clause? Did that really fix your problem of cascading down the updates on related objects in the .save call?

**Comment by Tidiane0843** [comment]
Euh yes. This is my code 
```
async update(newuser: User) {
       // newUser and My old user have the same matricule
       const getUserFromDB = await this.userRepository.findOne({matricule: newuser.matricule });
       this.userRepository.merge(getUserFromDB, newuser);
       this.userRepository.save(getUserFromDB)
          .then(
                      (data) => {
                          console.log('success');
                      },
                      (err) => {
                          // error
                          console.log(err);
                      },
                  );
    }
```
After adding ```
@ManyToOne(type => User, user=> user.photo, 
{onUpdate: 'CASCADE', onDelete: 'CASCADE'}
)
```
I have no more the error : EntityColumnNotFound: No entity column "photos" was found.
Hope it helps

**Comment by cristianvena18** [comment]
I have the same problem, but I can't use the save method since I have a unique column and it throws me an error if I don't update that column, any solution to this?

**Comment by bhubr** [comment]
To those stumbling on this issue: I think it has to do with _how to use_ TypeORM.

I have a `Task` entity, and each `Task` can have many `Timebox`es. So I had the same problem as mentioned here when I was attempting to update a `Task`:

    EntityColumnNotFound: No entity column "timeboxes" was found.

The problem was that my frontend was sending the `Task`'s data to the backend **with** `timeboxes`, the array of `Timebox`es. My payload was like this (simplified a bit):

```json
{
  "id": 31,
  "title": "Fix that nasty bug",
  "timeboxes": [{ "id": 132 }, { "id": 187 }]
}
```

when it should have been only the `Task` entity's intrinsic fields:

```json
{
  "id": 31,
  "title": "Fix that nasty bug"
}
```

**Comment by shaniqwa** [comment]
spent some time in this... couldn't get this to work using `update`. @TidianeRolland thank you, merge & save did the trick.

**Comment by sippeangelo** [comment]
So how are you actually supposed update a relation without fetching the whole entity with find, merge and save? This isn't atomic in any way. I don't want to update the whole entity data, I just want to add a row to the id-to-id table that describes the relation...

**Comment by allandiego** [comment]
> 
> 
> So how are you actually supposed update a relation without fetching the whole entity with find, merge and save? This isn't atomic in any way. I don't want to update the whole entity data, I just want to add a row to the id-to-id table that describes the relation...

I agree, i was able to fix this deleting all relations keys from object returned by find() and update works

**Comment by bestickley** [comment]
@sippeangelo, to atomically update, create a partial version of whatever you're trying to update, delete the relations off the entity, then update the partial object. Something like this:
```
const response = Object.assign(new Response(), {
      ...responseDto,
      completionDate,
    });
const partialResponse: Partial<Response> = { ...response };
    delete partialResponse.maapsResponse;
    await Promise.all([
      this.entityManager.update(
        MaapsResponse,
        { id: oldResponse.maapsResponse.id },
        maapsResponse
      ),
      this.entityManager.update(Response, { id: response.id }, partialResponse),
    ]);
```

**Comment by bestickley** [comment]
This really should be fixed. I see it's labelled with `driver:postgres`. I'm using SQL Server and still getting the error. One should be allowed to simply update the entity they're passing and the update function ignore or print out a warning that the relations won't be updated. It would also be helpful if it ignored the id column.

**Comment by GarryOne** [comment]
You should set `{ cascade: true}` on the parent relation (the one through you want to make the update) and ` { onDelete: true, onUpdate: true }` on the child relation. 

Also, note the differences between the `.save()` and `.update()` when working with relationships. 
```
/**
 * Updates entity partially. Entity can be found by a given conditions.
 * Unlike save method executes a primitive operation without cascades, relations and other operations included.
 * Executes fast and efficient UPDATE query.
 * Does not check if entity exist in the database.
 */
update(criteria: string | string[] | number | number[] | Date | Date[] | ObjectID | ObjectID[] | FindConditions<Entity>, partialEntity: QueryDeepPartialEntity<Entity>): Promise<UpdateResult>;
```
Therefore, I recommend using `.save()` instead of `.update()`.

**Comment by ZhaoKunLong** [comment]
> Euh yes. This is my code
> 
> ```
> async update(newuser: User) {
>        // newUser and My old user have the same matricule
>        const getUserFromDB = await this.userRepository.findOne({matricule: newuser.matricule });
>        this.userRepository.merge(getUserFromDB, newuser);
>        this.userRepository.save(getUserFromDB)
>           .then(
>                       (data) => {
>                           console.log('success');
>                       },
>                       (err) => {
>                           // error
>                           console.log(err);
>                       },
>                   );
>     }
> ```
> 
> After adding
> 
> ```
> @ManyToOne(type => User, user=> user.photo, 
> {onUpdate: 'CASCADE', onDelete: 'CASCADE'}
> )
> ```
> 
> I have no more the error : EntityColumnNotFound: No entity column "photos" was found.
> Hope it helps



> Euh yes. This is my code
> 
> ```
> async update(newuser: User) {
>        // newUser and My old user have the same matricule
>        const getUserFromDB = await this.userRepository.findOne({matricule: newuser.matricule });
>        this.userRepository.merge(getUserFromDB, newuser);
>        this.userRepository.save(getUserFromDB)
>           .then(
>                       (data) => {
>                           console.log('success');
>                       },
>                       (err) => {
>                           // error
>                           console.log(err);
>                       },
>                   );
>     }
> ```
> 
> After adding
> 
> ```
> @ManyToOne(type => User, user=> user.photo, 
> {onUpdate: 'CASCADE', onDelete: 'CASCADE'}
> )
> ```
> 
> I have no more the error : EntityColumnNotFound: No entity column "photos" was found.
> Hope it helps

Yeah! I also do like that. But every time will generate a new record, not update previous data. Do you meet this problem?

**Comment by ZhaoKunLong** [comment]
@TidianeRolland  Yeah! I also do like you do. But every time will generate a new record, not update previous data. Do you meet this problem?

**Comment by ZhaoKunLong** [comment]
I find a way to resolve this problem.can user `.save(entities, options)` to fix it.  
Tip: can not use `.save(entity, options)`  I don't why.

I used Typescript and NestJS

Do like this:

```
const images = [{id:1},{id:2},{id:3}];

const user = {
id: 1 // already exists
images //  save the images relation with user, it will generate may record in `user_image` table
}
await this.usersService.saveMultiple([ user ]); // why I use array. Cause save(entity,options) is not work
 
function  saveMultiple(entities: Array<User>, options?: SaveOptions): Promise<User[]> {
    try {
      return this.userRepository.save(entities, options);
    } catch (err) {
      throw new InternalServerErrorException();
    }
  }
```

Boom! it will work well.

I hope it will be of some help to you
